### PR TITLE
Proposal for upgrade API versions of K8s resources

### PR DESCRIPTION
**Pull request by michal-hudy**
**Description**

Changes proposed in this pull request:

- Proposal for upgrade API versions of K8s resources

**Resolves**

this PR resolves kyma-project/kyma#5546

## Decision log

| Name | Description |
|-----------------------|------------------------------------------------------------------------------------|
| Title | Proposal for upgrade API versions of K8s resources |
| Due date | 2019-09-30 |
| Status | `Proposed` |
| Decision type | `Binary` |
| Affected decisions | none |

## Context

mentioned in proposal

## Decision

mentioned in proposal

## Consequences

Temporary downtime of Kyma

### CODE REVIEW

File: collaboration/sig-core/proposals/upgrade-apiversion-of-k8s-resources.md [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+Created on 2019-09-18 by Michal Hudy (@michal-hudy).
+
+## Motivation
+
+API Versions of Kubernetes resources are changing in time and some of them are going to be deprecated. Later, deprecated versions are no longer served by API server. Currently, we have such situation with Kubernetes 1.16 (https://kubernetes.io/blog/2019/07/18/api-deprecations-in-1-16), where some API versions that we are using in Kyma will be no longer served. The problem with an upgrade is that field `apiVersion` is immutable, so the resource cannot be just updated, they needs to be recreated.
+
+## Solution
+
+Solution is not easy as Kubernetes cluster should work without downtime so in ideal world we should create a new resources next to the old one that are deprecated and when new resources are ready, then remove the old one. Unfortunately, world is not ideal and not all resources can be duplicated - ingresses, resources that expects only one instance in the cluster, etc.
```
Comment (mmitoraj) [inline collaboration/sig-core/proposals/upgrade-apiversion-of-k8s-resources.md]:
```suggestion
There is no perfect solution. We know that a Kubernetes cluster should work without downtime. In the ideal world, we should create new resources next to the deprecated ones, and when the new resources are ready then we should remove the old ones. Unfortunately, not all resources can be duplicated, for example, ingresses or resources that expect only one instance in the cluster.
```

File: collaboration/sig-core/proposals/upgrade-apiversion-of-k8s-resources.md [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+In such a case the downtime will be necessary, and we need to the following:
```
Comment (mmitoraj) [inline collaboration/sig-core/proposals/upgrade-apiversion-of-k8s-resources.md]:
```suggestion
In such a case the downtime will be necessary, and we need to do the following:
```

File: collaboration/sig-core/proposals/upgrade-apiversion-of-k8s-resources.md [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+- Create a `pre-upgrade` Helm hook that will be removing old resource definition if there is a new API version available in the cluster
+- Update API versions in resources if new API version is available.
```
Comment (mmitoraj) [inline collaboration/sig-core/proposals/upgrade-apiversion-of-k8s-resources.md]:
```suggestion
- Update API versions in resources if a new API version is available.
```

File: collaboration/sig-core/proposals/upgrade-apiversion-of-k8s-resources.md [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+Pre-upgrade hook may look as following:
```
Comment (mmitoraj) [inline collaboration/sig-core/proposals/upgrade-apiversion-of-k8s-resources.md]:
```suggestion
See the following example of how the `pre-upgrade hook` may look:
```

**Comment by michal-hudy** [comment]
@k15r thanks for checking it! I tested it with RC release of k8s 1.16 and kubctl 1.14.3. I will check it again

**Comment by mszostok** [comment]
@derberg exactly, I had the same question. Basically The K8s project maintains release branches for the most recent three minor releases, so there is even no point in having compatibility with version below 1.9. IMO we should drop the template and just update the apiVersion.

**Comment by michal-hudy** [comment]
Ok guys, I agree. I will update the proposal, it will be even simpler? Do we even need it as we already have an issue about updating apiVersion?

**Comment by derberg** [comment]
proposal still makes sense as that was a work and investigation that had to be done, and if outcome of proposal is just to make sure we use latest version of all tools, it is still valid proposal :)

File: collaboration/sig-core/proposals/upgrade-apiversion-of-k8s-resources.md [hunk 5]
```diff
@@ -0,0 +1,23 @@
+# Upgrade API versions of Kubernetes resources
+
+Created on 2019-09-18 by Michal Hudy (@michal-hudy).
+
+## Motivation
+
+API Versions of Kubernetes resources are changing in time and some of them are going to be deprecated. Later, deprecated versions are no longer served by the API server. Currently, we have such a situation with Kubernetes 1.16 (https://kubernetes.io/blog/2019/07/18/api-deprecations-in-1-16), where some API versions that we are using in Kyma will be no longer served. For now, Kyma is not compatible with Kubernetes 1.16.
```
Comment (mmitoraj) [inline collaboration/sig-core/proposals/upgrade-apiversion-of-k8s-resources.md]:
```suggestion
API Versions of Kubernetes resources change with time and some of them become deprecated. Once deprecated, the versions are no longer served by an API server. Currently, we have such a situation with [Kubernetes 1.16](https://kubernetes.io/blog/2019/07/18/api-deprecations-in-1-16), where some API versions that we use in Kyma will be no longer served. For now, Kyma is not compatible with Kubernetes 1.16.
```

File: collaboration/sig-core/proposals/upgrade-apiversion-of-k8s-resources.md [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+## Test Environment
+
+I have tested proposed solution with:
```
Comment (mmitoraj) [inline collaboration/sig-core/proposals/upgrade-apiversion-of-k8s-resources.md]:
```suggestion
I tested the proposed solution with:
```

File: collaboration/sig-core/proposals/upgrade-apiversion-of-k8s-resources.md [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+ - kubectl v1.16.1
+ - helm v2.14.3
+
+## Solution
+
+It looks like a solution is quite easy, but it is important to use the newest tools. First,  we need to update `apiVersion` field in charts and apply such changes to the cluster. The next step is changing the version in code, to be sure that our applications are asking and operating on the latest `apiVersion`. Later Kubernetes version can be upgraded.
```
Comment (mmitoraj) [inline collaboration/sig-core/proposals/upgrade-apiversion-of-k8s-resources.md]:
```suggestion
The solution is quite easy, but it is important to use the latest tools. First,  we need to update the `apiVersion` field in charts and apply such changes to the cluster. The next step is changing the version in code to be sure that our applications call and operate on the latest `apiVersion`. Only then the Kubernetes version can be upgraded.
```

File: collaboration/sig-core/proposals/upgrade-apiversion-of-k8s-resources.md [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+## Action items
+
+* Update `apiVersions` to the newest one
+* Update code to operate on the newest `apiVersions`
+* Pipelines are using newest version of `kubectl` or at least the same as Kubernetes version
```
Comment (mmitoraj) [inline collaboration/sig-core/proposals/upgrade-apiversion-of-k8s-resources.md]:
```suggestion
* Run pipelines using the latest version of `kubectl` or at least the same as the Kubernetes version
```

File: collaboration/sig-core/proposals/upgrade-apiversion-of-k8s-resources.md [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+ - kubectl v1.16.1
+ - helm v2.14.3
+
+## Solution
+
+It looks like a solution is quite easy, but it is important to use the newest tools. First,  we need to update `apiVersion` field in charts and apply such changes to the cluster. The next step is changing the version in code, to be sure that our applications are asking and operating on the latest `apiVersion`. Later Kubernetes version can be upgraded.
+
+## Action items
+
+* Update `apiVersions` to the newest one
```
Comment (mmitoraj) [inline collaboration/sig-core/proposals/upgrade-apiversion-of-k8s-resources.md]:
```suggestion
* Update `apiVersions` to the latest version
```

File: collaboration/sig-core/proposals/upgrade-apiversion-of-k8s-resources.md [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+* Update code to operate on the newest `apiVersions`
```
Comment (mmitoraj) [inline collaboration/sig-core/proposals/upgrade-apiversion-of-k8s-resources.md]:
```suggestion
* Update code to operate on the latest `apiVersions`
```

**Pull request merged**